File: swift/Sources/TokenTrackerCore/Scanners/PiScanner.swift
//
//  PiScanner.swift
//  TokenTrackerCore
//
//  移植自 scanners/pi.py：~/.pi/agent/sessions/**/*.jsonl（Oh My Pi: ~/.omp）。
//  幂等键：文件内事件 id；官方 cost 优先，缺失时价格表估算。
//

import Foundation

public struct PiScanner: ScannerAdapter {
    public let name = "pi"
    public let detail = "~/.pi/agent/sessions/**/*.jsonl"
    public let roots: [String]

    public init(roots: [String]) {
        self.roots = roots.map(expandPath).filter {
            var isDir: ObjCBool = false
            return FileManager.default.fileExists(atPath: $0, isDirectory: &isDir) && isDir.boolValue
        }
    }

    public func detect() -> Bool {
        !roots.isEmpty
    }

    private func parseTS(_ raw: Any?) -> Int64 {
        if let n = raw as? NSNumber, CFGetTypeID(n) != CFBooleanGetTypeID() {
            let d = n.doubleValue
            return Int64(d < 1e12 ? d * 1000 : d)
        }
        if let s = raw as? String { return parseISODateMs(s) ?? 0 }
        return 0
    }

    public func scan(_ store: UsageStore, _ prices: PriceTable, full: Bool) throws -> ScanOutcome {
        var cursor = try store.getScanCursor(tool: name)
        let effectiveFull = full || activityNeedsBackfill(cursor)
        var outcome = ScanOutcome()
        for base in roots {
            var files: [String] = []
            if let enumerator = FileManager.default.enumerator(atPath: base) {
                for case let entry as String in enumerator where entry.hasSuffix(".jsonl") {
                    files.append((base as NSString).appendingPathComponent(entry))
                }
            }
            files.sort()
            for path in files {
                if !effectiveFull && !fingerprintChanged(cursor: cursor, path: path) { continue }
                guard let statKey = StatKey(path: path) else { continue }
                outcome.files += 1
                var sessionID = ""
                var project = ((path as NSString).deletingLastPathComponent as NSString)
                    .lastPathComponent
                var title: String?
                for (lineno, obj) in iterJSONL(path) {
                    if title == nil {
                        let text = userText(obj)
                        if !text.isEmpty { title = text }
                    }
                    let type = obj["type"] as? String
                    if type == "session" {
                        sessionID = obj["id"] as? String ?? ""
                        let cwd = obj["cwd"] as? String ?? ""
                        if !cwd.isEmpty { project = cwd }
                        continue
                    }
                    guard type == "message",
                          let msg = obj["message"] as? [String: Any] else { continue }
                    let ts = parseTS(jsonOrAny(msg["timestamp"], obj["timestamp"]))
                    for (index, value) in (msg["content"] as? [Any] ?? []).enumerated() {
                        guard let part = value as? [String: Any],
                              let partType = part["type"] as? String else { continue }
                        if partType == "toolCall", let rawName = part["name"] as? String {
                            let callID = jsonOrString(part["id"], part["toolCallId"])
                            let fallback = "\(obj["id"].map(String.init(describing:)) ?? String(lineno))|\(index)"
                            let change = try store.recordActivity(
                                agent: name, srcKey: "\(realPath(path))|tool|\(callID.isEmpty ? fallback : callID)",
                                rawName: rawName, sessionID: sessionID, callID: callID,
                                startedAt: ts == 0 ? nil : ts, sourceKind: "pi_jsonl",
                                arguments: jsonOrAny(part["arguments"], part["input"]))
                            outcome.activityAdded += change.added
                            outcome.activityUpdated += change.updated
                        } else if partType == "toolResult" {
                            var status = (part["isError"] as? Bool) == true
                                ? "error" : ActivityNormalizer.status(part["status"])
                            if status == "unknown" { status = "success" }
                            outcome.activityUpdated += try store.completeActivity(
                                agent: name, callID: jsonOrString(part["toolCallId"], part["id"]),
                                status: status, endedAt: ts == 0 ? nil : ts)
                        }
                    }
                    guard let usage = msg["usage"] as? [String: Any] else { continue }
                    let inp = jsonInt(usage["input"])
                    let outp = jsonInt(usage["output"])
                    let cr = jsonInt(usage["cacheRead"])
                    let cw = jsonInt(usage["cacheWrite"])
                    if inp + outp + cr + cw == 0 { continue }
                    let model = jsonOrString(msg["model"], obj["modelId"])
                    let basename = (path as NSString).lastPathComponent
                    let eventID = (obj["id"] as? NSNumber)?.stringValue
                        ?? (obj["id"] as? String) ?? "None"
                    let key = "\(basename)|\(eventID)"
                    let costObj = usage["cost"] as? [String: Any] ?? [:]
                    // 官方 cost 优先；<=0 时价格表估算（未匹配保持 nil，不计费）
                    let sourceCost = (costObj["total"] as? NSNumber)?.doubleValue ?? 0
                    let sourceProvider = jsonOrString(msg["provider"], obj["provider"])
                    let quote = prices.quote(
                        provider: sourceProvider.isEmpty ? nil : sourceProvider,
                        model: model, input: inp, output: outp, cacheRead: cr,
                        cacheWrite: cw, eventAtMs: ts)
                    let cost = sourceCost > 0 ? sourceCost : quote?.cost
                    outcome.added += try store.putEvent(
                        tool: name, srcKey: key, sessionID: sessionID, project: project,
                        ts: ts, model: model, input: inp, output: outp,
                        cacheRead: cr, cacheWrite: cw, cost: cost,
                        costSource: sourceCost > 0 ? "native" : "estimate",
                        provider: quote?.provider ?? sourceProvider,
                        priceVersionID: sourceCost > 0 ? nil : quote?.priceVersionID)
                }
                cursor[path] = statKey.asDict
                if let title {
                    try store.setSessionTitle(tool: name, sessionID: sessionID, title: title)
                }
            }
        }
        markActivityCurrent(&cursor)
        try store.setScanCursor(tool: name, cursor: cursor)
        return outcome
    }
}

File: windows/src/TokenTracker.Core/Scanners/PiScanner.cs
using TokenTracker.Core.Activity;
using TokenTracker.Core.Json;
using TokenTracker.Core.Platform;
using TokenTracker.Core.Pricing;
using TokenTracker.Core.Store;

namespace TokenTracker.Core.Scanners;

/// <summary>
/// 移植 PiScanner.swift（scanners/pi.py）：~/.pi/agent/sessions/**/*.jsonl（Oh My Pi: ~/.omp）。
/// 幂等键：文件内事件 id；官方 cost 优先，缺失时价格表估算。
/// </summary>
public sealed class PiScanner : IScanner
{
    public string Name => "pi";
    public string Detail => "~/.pi/agent/sessions/**/*.jsonl";
    public IReadOnlyList<string> Roots { get; }

    public PiScanner(IEnumerable<string> roots) =>
        Roots = roots.Select(WinPaths.Expand).Where(Directory.Exists).ToList();

    public bool Detect() => Roots.Count > 0;

    public ScanOutcome Scan(UsageStore store, PriceTable prices, bool full)
    {
        var cursor = store.GetScanCursor(Name);
        var effectiveFull = full || ActivityNormalizer.NeedsBackfill(cursor);
        var outcome = new ScanOutcome();
        foreach (var root in Roots)
        {
            var files = WinPaths.EnumerateFilesRelative(root, rel => rel.EndsWith(".jsonl", StringComparison.Ordinal))
                .Select(rel => WinPaths.JoinRelative(root, rel));
            foreach (var path in files)
            {
                if (!effectiveFull && !FileIdentity.Changed(cursor, path)) continue;
                if (FileIdentity.Of(path) is not { } statKey) continue;
                outcome.Files++;
                var sessionId = "";
                var project = WinPaths.LastComponent(WinPaths.Parent(path));
                string? title = null;
                string? realPath = null;
                foreach (var (_, obj) in ScannerSupport.IterJsonl(path))
                {
                    if (title is null)
                    {
                        var text = ScannerSupport.UserText(obj);
                        if (text.Length > 0) title = text;
                    }
                    var type = obj.Get("type") as string;
                    if (type == "session")
                    {
                        sessionId = obj.Get("id") as string ?? "";
                        if (obj.Get("cwd") is string { Length: > 0 } cwd) project = cwd;
                        continue;
                    }
                    if (type != "message" || obj.Get("message") is not Dictionary<string, object?> msg) continue;
                    var ts = ScannerSupport.ParseTs(PyJson.Or(msg.Get("timestamp"), obj.Get("timestamp")));
                    if (msg.Get("content") is List<object?> content)
                    {
                        for (var index = 0; index < content.Count; index++)
                        {
                            if (content[index] is not Dictionary<string, object?> part
                                || part.Get("type") is not string partType) continue;
                            if (partType == "toolCall" && part.Get("name") is string rawName)
                            {
                                var callId = PyJson.OrString(part.Get("id"), part.Get("toolCallId"));
                                // Python: call_id or str(obj.get('id')) + '|' + str(index)
                                var fallback = $"{PyJson.PyStr(obj.Get("id"))}|{index}";
                                realPath ??= WinPaths.Real(path);
                                var change = store.RecordActivity(Name,
                                    $"{realPath}|tool|{(callId.Length == 0 ? fallback : callId)}", rawName, sessionId,
                                    callId: callId, startedAt: ts == 0 ? null : ts, sourceKind: "pi_jsonl",
                                    arguments: PyJson.Or(part.Get("arguments"), part.Get("input")));
                                outcome.ActivityAdded += change.Added;
                                outcome.ActivityUpdated += change.Updated;
                            }
                            else if (partType == "toolResult")
                            {
                                var status = part.Get("isError") is true ? "error" : ActivityNormalizer.Status(part.Get("status"));
                                if (status == "unknown") status = "success";
                                outcome.ActivityUpdated += store.CompleteActivity(Name,
                                    PyJson.OrString(part.Get("toolCallId"), part.Get("id")), status, ts == 0 ? null : ts);
                            }
                        }
                    }
                    if (msg.GetDict("usage") is not { } usage) continue;
                    var inp = PyJson.JInt(usage.Get("input"));
                    var outp = PyJson.JInt(usage.Get("output"));
                    var cr = PyJson.JInt(usage.Get("cacheRead"));
                    var cw = PyJson.JInt(usage.Get("cacheWrite"));
                    if (inp + outp + cr + cw == 0) continue;
                    var model = PyJson.OrString(msg.Get("model"), obj.Get("modelId"));
                    var eventId = obj.Get("id") switch
                    {
                        string s => s,
                        long or double => PyJson.PyStr(obj["id"]),
                        _ => "None",
                    };
                    var key = $"{WinPaths.LastComponent(path)}|{eventId}";
                    // 官方 cost 优先；<=0 时价格表估算（未匹配保持 null，不计费）
                    double? cost = PyJson.AsDouble(usage.GetDict("cost").Get("total")) ?? 0;
                    if (cost <= 0) cost = prices.Cost(model, inp, outp, cr, cw);
                    outcome.Added += store.PutEvent(Name, key, sessionId, project, ts, model, inp, outp, cr, cw, cost);
                }
                cursor[path] = statKey.AsDict();
                if (title is not null) store.SetSessionTitle(Name, sessionId, title);
            }
        }
        ActivityNormalizer.MarkCurrent(cursor);
        store.SetScanCursor(Name, cursor);
        return outcome;
    }
}
